frontend/class-handle-404.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * WPSEO plugin file.
+ *
+ * @package WPSEO\Frontend
+ */
+
+/**
+ * Class WPSEO_Handle_404
+ *
+ * Handles intercepting requests.
+ *
+ * @since 9.4
+ */
+class WPSEO_Handle_404 implements WPSEO_WordPress_Integration {
+
+	/**
+	 * Registers all hooks to WordPress.
+	 *
+	 * @return void
+	 */
+	public function register_hooks() {
+		add_filter( 'pre_handle_404', array( $this, 'handle_404' ) );
+	}
+
+	/**
+	 * Handle the 404 status code.
+	 *
+	 * @param bool $handled Whether we've handled the request.
+	 *
+	 * @return bool True if it's 404.
+	 */
+	public function handle_404( $handled ) {
+
+		if ( is_feed() ) {
+			return $this->is_feed_404( $handled );
+		}
+
+		return $handled;
+	}
+
+	/**
+	 * If there are no posts in a feed, make it 404 instead of sending an empty RSS feed.
+	 *
+	 * @global WP_Query $wp_query
+	 *
+	 * @param bool $handled Whether we've handled the request.
+	 *
+	 * @return bool True if it's 404.
+	 */
+	private function is_feed_404( $handled ) {
+		global $wp_query;
+
+		// Don't 404 if the query contains posts or if it matched an object.
+		if ( $wp_query->posts || get_queried_object() || is_home() ) {
```

Comment:
Please use the `$wp_query` methods for `get_queried_object ` and `is_home` because that global variable is already used here.

Reply:
👍